internal/command/local/disconnect.go

```diff
@@ -31,20 +36,93 @@ func runDisconnect(cfg *config.LocalDisconnect, args []string) error {
-	if _, err = rootManagerClient.Shutdown(ctx, &rmapi.ShutdownRequest{}); err != nil {
-		return fmt.Errorf("error requesting shutdown in sandbox manager api: %v", err)
+
+	runState := &runState{}
+	ticker := time.NewTicker(time.Second / 10)
+	defer ticker.Stop()
+	wasRunning := false
+	for !runState.isDone() {
+		wasRunning = runState.init(signadotDir) || wasRunning
+		err := runState.tryKill()
+		_ = err
+		<-ticker.C
+	}
+
+	if wasRunning {
+		fmt.Printf("disconnected.\n")
+		return nil
+	}
+	return fmt.Errorf("signadot was not connected")
+}
+
+// process monitor result for one of rootmanager, sandbox manager
+type mgrRunState struct {
+	C   <-chan struct{}
+	PID int
+	E   error
+}
+
+// we have a sandbox manager and a root manager to stop, and may be
+// run in unprivileged mode, without root manager, so
+// for stopping them we keep track of both.
+type runState struct {
+	Root    mgrRunState
+	NotRoot mgrRunState
+}
+
+// we are done when there are no more pid files
+func (rs *runState) isDone() bool {
+	return os.IsNotExist(rs.Root.E) && os.IsNotExist(rs.NotRoot.E)
+}
+
+func (rs *runState) tryKill() error {
+	if rs.Root.C != nil {
+		// Establish a connection with sandbox manager
+		grpcConn, err := grpc.Dial("127.0.0.1:6667", grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return fmt.Errorf("couldn't connect sandbox manager api, %w", err)
+		}
+		defer grpcConn.Close()
+
+		// Send the shutdown order
+		rootManagerClient := rmapi.NewRootManagerAPIClient(grpcConn)
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if _, err = rootManagerClient.Shutdown(ctx, &rmapi.ShutdownRequest{}); err != nil {
+			return fmt.Errorf("error requesting shutdown in root manager api: %w", err)
+		}
+		return nil
+	}
+	if rs.NotRoot.C != nil {
+		// Establish a connection with root manager
+		grpcConn, err := grpc.Dial("127.0.0.1:6666", grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return fmt.Errorf("couldn't connect sandbox manager api, %w", err)
+		}
+		defer grpcConn.Close()
+
+		// Send the shutdown order
+		sbManagerClient := sbmapi.NewSandboxManagerAPIClient(grpcConn)
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if _, err := sbManagerClient.Shutdown(ctx, &sbmapi.ShutdownRequest{}); err != nil {
+			return fmt.Errorf("error requesting shutdown in sandbox manager api: %w", err)
+		}
 	}
-	// TODO: monitor shutdown works
 	return nil
 }
+
+func (rs *runState) init(signadotDir string) bool {
+	rmPIDFile := filepath.Join(signadotDir, config.RootManagerPIDFile)
+	sbmPIDFile := filepath.Join(signadotDir, config.SandboxManagerPIDFile)
+	processes.CleanPIDFile(rmPIDFile)
+	processes.CleanPIDFile(sbmPIDFile)
+
+	rs.Root.C, rs.Root.PID, rs.Root.E = processes.MonitorPID(rmPIDFile, slog.Default())
+	rs.NotRoot.C, rs.NotRoot.PID, rs.NotRoot.E = processes.MonitorPID(sbmPIDFile, slog.Default())
```

Comment:
As far as I understand, during a disconnect process, we will be calling several times this function, and every time we call `processes.MonitorPID()` that's launching a new go runtime that will be running in background until the process ends. So, we would end up creating many goroutines here, am I right?

Reply:
yes I guess so.  thinking about a solution.

Reply:
updated